### DISCUSSION TITLE
ESP32 WROOM with ILI9486 parallel tft shield

### DISCUSSION THREAD

**Discussion by kiduino** [Bodmer/TFT_eSPI#2154] [Q&A - General]
hi, I have a 320x480 3.5" ili9486 display, i beleive that it is probably made to fit uno/mega boards? 
![61jJ4fdTBML _AC_SL1500_](https://user-images.githubusercontent.com/81905897/200637355-50a106bf-a3fe-4c0e-b7ad-52960ebc25ab.jpg)

im using an esp32 wroom(not an esp32 UNO board). I cant find any posts that describe using an esp32 WROOM to get touch data from my tft, only ESP32 UNO board managing to get touch data.

is there any way of getting touch data from the ili9486 tft to my esp32 wroom? many similar discussions suggest using [this library](https://github.com/s60sc/Adafruit_TouchScreen) however i belive that library only works with esp32 uno boards not devkits like mine

**Comment by Bodmer** [comment]
The UNO format of the ESP32 boards are still available so that is the simplest route. There's nothing special about the capability of the UNO style ESP32 boards so other board types could be used with the same or equivalent GPIO allocations.

I have not used the touch library but it should work fine with the ESP32 providing the required number of GPIO are available.

**Reply by kiduino** [reply]
hey Bodmer, my esp32 board has 34 gpios, and im glad to hear that its possible. Ive installed the touch library but unfortuanately it doesnt seem to be working, i believe this could be because ive gotten the tft touch pins wrong. Would you happen to know which pins on the 3.2" ili9486 parallel display are for touch xp, touch yp, touch xm and touch ym??

**Reply by Bodmer** [reply]
It is tricky to work out because the TFT shield uses pins for dual purposes and so to get the shield to work with the UNO format ESP32 board a number of links had to be made so that some pins can be used for both digital and analoge functions.

This means that some "shield pins" are connected to two processor GPIO pins.

This is the connections I think you need to copy the UNO style connections to the shield:
```
TFT_CS    33  and 34      // Chip select control pin and touch interface
TFT_DC   15  and 35      // Data Command control pin and touch interface
TFT_RST  32  and 36      // Reset pin and touch interface pin

TFT_WR   4  and 39    // Write strobe control pin and touch interface pin
TFT_RD    2

TFT_D0   12      // Must use pins in the range 0-31 for the data bus
TFT_D1   13      // so a single register write sets/clears all bits
TFT_D2   26
TFT_D3   25
TFT_D4   17
TFT_D5   16
TFT_D6   27
TFT_D7   14
```
Clearly you also have to connect power and 0v (GND) to the shield too.

**Reply by kiduino** [reply]
thanks! this is very helpful. ill start wiring it up and see if it works

**Reply by kiduino** [reply]
hello again! ive wired my tft to my esp32 using the pins you sent, and unfortuanately i cant get touch data, only display stuff, even after checking my pins multiple times.  seeing as my wiring is correct, could the pin definitions im using for the touchscreen example be wrong/need changing ?
![Screenshot 2022-11-09 191700](https://user-images.githubusercontent.com/81905897/200921095-cde0c585-2e8e-41c8-bd7e-b96ecaba8b29.png)
or these pins inside of the header file:
```
// ESP32 specific 
#define ESP32_WIFI_TOUCH // uncomment to use parallel MCU Friend LCD touchscreen with ESP32 UNO Wifi
#ifdef ESP32 
#define ADC_MAX 4095  // maximum value for ESP32 ADC (default 11db, 12 bits)
#define aXM 35  // analog input pin connected to LCD_RS (DC)
#define aYP 39  // analog input pin connected to LCD_WR
```

**Reply by Bodmer** [reply]
If you are using the latest ESP32 board package I think that ESP32 is not defined. So modify TouchScreen.h. Change this line:
```
#ifdef ESP32
```
to
```
#ifdef ESP32_WIFI_TOUCH
```

**Reply by kiduino** [reply]
no better result from that im afraid
a little side note: the reason i havent bought a esp32 uno style board is becuase my project needs to be as compact as posibble, and the uno board is very bulky and probably would leave me no space for a battery for my handheld device. i cant buy any spi tft displays as the closest i could find would take atleast a month to arrive.

**Reply by Bodmer** [reply]
I can't see where the problem is. As far as I can tell it should work, all the pins look good. Post a link to the display you are using.

The MCUfriend library should also run with those pin definitions and has some diagnostic sketches for the touch interface but it is not a library I am familiar with. David Prentice is very good at helping out on the Arduino forum though.

See here:
https://forum.arduino.cc/t/mcufriend-2-4-tft-pinout-where-are-the-touchscreen-signal-pins/496361/8

**Reply by Bodmer** [reply]
Library mentioned is here:
https://github.com/prenticedavid/MCUFRIEND_kbv

**Reply by kiduino** [reply]
https://www.amazon.co.uk/ALMOCN-digitizer-Compatible-Mega2560-Raspberry/dp/B094CGQ426/ref=sr_1_6?crid=8FI3ERA5VMLZ&keywords=3.5%22+tft&qid=1667929251&sprefix=3.5+tft%2Caps%2C97&sr=8-6
the exact tft im using

**Reply by kiduino** [reply]
come up with another possibility why its not working: my esp32 wroom board is really weird. it could just be printing errors, but pin 33 on the board is labelled 23, meaning that it looks like the board has two 23 pins. another thing that could be a printing error is that pin 36 is labelled "SP" and pin 39 is labelled "SM". i did some diagnosis on these pins to see if they work properly, starting by setting pin 36 and 39 as INPUT, then a simple sketch that prints "39 high" if 39 is digital read high or prints "36 high" if pin 36 is high.
when i attached 5v pin 39 it would read high, but occasional 36 would read high aswell. could this just be because the pins are very sensetive? or are they linked in some way. this combined with their weird pin labelling seems very odd

**Reply by kiduino** [reply]
actually im goint to guess its just because the pins are extremely sensetive as they can detect my finger as HIGH when i touch them

**Comment by kiduino** [comment]
hello, upon more examination i have found the issue that is stopping everything from working. 
the display only works if pin 15 and 35 are NOT wired together. all the other wiring that you showed works fine.
if i wire 15 and 35 together, the display simply shows white and does not run any examples, any ideas on how to work around this, could a different pin be used? edit:
it also happens when wiring pin 4 to pin 39
